### ISSUE TITLE
parser: flagged subscript pattern with a bracket expression rejected in arithmetic context

### ISSUE THREAD

**Issue by ss-o** [z-shell/zsh-lint#368]
## Summary

A flagged subscript whose pattern holds a bracket expression before `#` is rejected when the subscript stands in an arithmetic expression.
Native Zsh accepts it.

```zsh
print $(( m[(r)a[^:]##] ))
print ${$(( m[(r)a[^:]##] ))}
```

```console
$ zsh -f -n repro.zsh; echo $?
0
$ zsh-lint-survey repro.zsh
FAIL repro.zsh
repro.zsh:1:23: not a valid arithmetic operator: `#`
```

The same subscript without a bracket expression parses: `print ${$(( m[(r)abc] ))}` is `OK`.

## Scope

Found while verifying #283 (PR #367).
It is the arithmetic-context counterpart of #283: mvdan/sh ends the flagged pattern at the bracket expression's own `]`, and in arithmetic the `#` that follows is reported as an invalid arithmetic operator.

It is not caused by #367: the failure is byte-identical on `main` (`b2e5b48`) and on the PR head.
#367 does not cover it because its repairs gate on parameter-expansion errors and on a parse that succeeded, and neither fires in arithmetic.

## Evidence to gather before fixing

- Native semantics of a flagged pattern inside `$(( ))` and `(( ))`, verified by evaluation rather than `-n` alone.
- Whether the arithmetic subscript path shares `scanFlagPatternBrackets`, or needs its own gate on `not a valid arithmetic operator`.
- Composition with the nested-arithmetic resolver (#364).

**Comment by ss-o** [comment]
# Investigation notes for #368 (read-only; no implementation)

Recorded by the session that merged #375/#377, before a fresh session claims
the bug-368 task. Everything below is measured, not inferred.

Binary: built from merged `main` (`4c690f7`, i.e. after both #371 and #373
landed). Zsh 5.9.2.

## 1. It still reproduces after #375 changed scanFlagPatternBrackets

This was the open risk, since #368 was deferred precisely because #375 touched
the same function.

| Row | `zsh -f -n` | merged main |
| --- | --- | --- |
| `print $(( m[(r)a[^:]##] ))` | 0 | `1:21: not a valid arithmetic operator: \`#\`` |
| `print ${$(( m[(r)a[^:]##] ))}` | 0 | `1:23: not a valid arithmetic operator: \`#\`` |
| `print ${$(( m[(r)abc] ))}` (control, no bracket expr) | 0 | OK |

Unchanged from the issue's report. #375 neither fixed nor worsened it.

## 2. The construct genuinely evaluates (not just `-n` clean)

Per the #371 lesson that `zsh -n` alone cannot prove a row valid, each row was
RUN as a file with `typeset -a m; m=( 'a:1' 'abc' 'a:2' )`:

| Row | run rc | prints |
| --- | --- | --- |
| `print $(( m[(r)a[^:]##] ))` | 0 | `0` |
| `print ${$(( m[(r)a[^:]##] ))}` | 0 | `0` |
| `print $(( m[(i)a[^:]##] ))` | 0 | `4` |

The `(i)` row returning an index proves the subscript is really evaluated, so a
fixture should assert a VALUE, not merely parse. `(( m[(r)a[^:]##] ))` exits 1
because the arithmetic result is 0 — expected for an arithmetic command, not a
parse problem.

## 3. `#` in an arithmetic subscript is a REAL operator

This is the load-bearing finding for the design question the issue raises.

| Row | run | prints |
| --- | --- | --- |
| `print $(( 2#101 ))` | 0 | `5` |
| `print $(( m[2#10] ))` | 0 | `0` |
| `print $(( m[#] ))` | 0 | `0` |

`2#101` is radix notation. So `#` cannot be masked wholesale in an arithmetic
subscript, and **widening `isBareKeyPunctuation` in `subscript_compat.go` is the
wrong fix**: that set is documented as deliberately narrow so "arithmetic
subscripts on ordinary arrays keep their operators", and `#` is excluded for
exactly this reason.

Answering the issue's own question: the arithmetic subscript path already gates
on `not a valid arithmetic operator:` via `invalidSubscriptArithmetic`
(`subscript_compat.go:15`), but through `parseAssociativeSubscript`, whose mask
set excludes `#`. #368 therefore needs its own gate keyed on the flagged-pattern
shape, not an extension of the key-punctuation mask.

## 4. Suggested shape of the fix

The pattern is `name[(flags)PATTERN]` where PATTERN contains a bracket
expression, and mvdan/sh ends the flagged pattern at the bracket expression's
own `]` — the same root cause as #283/#371, but surfacing as an arithmetic
error because the remainder is read as arithmetic.

`scanFlagPatternBrackets` (`subscript_flag_pattern.go:295`) already computes the
correct pattern end by Zsh's bracket nesting and returns the edits needed. The
narrow fix is a new adapter that fires on `not a valid arithmetic operator:`
when a flagged subscript opener (`name[(flags)`) precedes the reported offset,
reuses `scanFlagPatternBrackets`, and masks only within the pattern it returns.
That keeps `2#101` untouched, because no flagged opener precedes it.

Gate precisely: the reported byte must sit inside the span
`scanFlagPatternBrackets` reports, not merely after a `[`.

## 5. Verification this fix will need

- Both oracles per row, and RUN every fixture (`(i)` rows assert a value).
- Radix rows (`2#101`, `m[2#10]`) in the must-keep-working set — a mask that
  eats them is a regression the corpus will not catch.
- The false-accept direction: generate `#`-in-arithmetic rows where Zsh REJECTS
  and confirm base and fixed both still reject. Per the #373 lesson, a probe set
  is only as good as its shape coverage — include radix, `(i)`/`(r)`/`(R)`
  flags, associative vs ordinary arrays, and `$(( ))` vs `(( ))` vs `${$(( ))}`.
- Composition with the nested-arithmetic resolver (#364), as the issue asks.
- Mutation-test the new gate, including a mutant that widens it to any `#`.
